### PR TITLE
ENH: DTI/TDI as_unit

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -108,6 +108,7 @@ Other enhancements
 - :meth:`DataFrame.plot.hist` now recognizes ``xlabel`` and ``ylabel`` arguments (:issue:`49793`)
 - Improved error message in :func:`to_datetime` for non-ISO8601 formats, informing users about the position of the first error (:issue:`50361`)
 - Improved error message when trying to align :class:`DataFrame` objects (for example, in :func:`DataFrame.compare`) to clarify that "identically labelled" refers to both index and columns (:issue:`50083`)
+- Added :meth:`DatetimeIndex.as_unit` and :meth:`TimedeltaIndex.as_unit` to convert to different resolutions; supported resolutions are "s", "ms", "us", and "ns" (:issue:`??`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v2.0.0.rst:111]:
```suggestion
- Added :meth:`DatetimeIndex.as_unit` and :meth:`TimedeltaIndex.as_unit` to convert to different resolutions; supported resolutions are "s", "ms", "us", and "ns" (:issue:`50616`)
```

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -396,6 +396,14 @@ class DatetimeTimedeltaMixin(DatetimeIndexOpsMixin):
 
     _join_precedence = 10
 
+    @property
+    def unit(self) -> str:
+        return self._data.unit
+
+    def as_unit(self: _TDT, unit: str) -> _TDT:
```
Comment (phofl) [inline pandas/core/indexes/datetimelike.py:403]: Can we somehow add a public docstring?

**Review by phofl — APPROVED** [review]

File: doc/source/whatsnew/v2.0.0.rst [hunk 3]
```diff
@@ -108,6 +108,7 @@ Other enhancements
 - :meth:`DataFrame.plot.hist` now recognizes ``xlabel`` and ``ylabel`` arguments (:issue:`49793`)
 - Improved error message in :func:`to_datetime` for non-ISO8601 formats, informing users about the position of the first error (:issue:`50361`)
 - Improved error message when trying to align :class:`DataFrame` objects (for example, in :func:`DataFrame.compare`) to clarify that "identically labelled" refers to both index and columns (:issue:`50083`)
+- Added :meth:`DatetimeIndex.as_unit` and :meth:`TimedeltaIndex.as_unit` to convert to different resolutions; supported resolutions are "s", "ms", "us", and "ns" (:issue:`50616`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.0.0.rst:111]: looks like other units work

e.g.
```python
In [24]: dti = date_range('2000', periods=10)

In [25]: dti.as_unit('D').unit
Out[25]: 'D'
```

should this raise?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Nice!

I just don't really follow the note on supported resolutions - should non-supported ones raise?

**Comment by jbrockmendel** [comment]
> I just don't really follow the note on supported resolutions - should non-supported ones raise?

good catch, will fix.

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me pending green, thanks @jbrockmendel !

**Review by mroeschke — APPROVED** [review]

**Pull request merged**